File: frontend/src/types/operations.ts
// operations domain types
import type { IntradaySourceStatusView } from "./common";
import type { PortfolioSummaryView } from "./portfolio";
import type { SimulationWorkspaceResponse } from "./simulation";
import type { RecommendationReplayView } from "./stock";
import type { ManualResearchRequestView } from "./research";

export interface OperationsOverviewView {
  generated_at: string;
  beta_readiness?: string;
  manual_portfolio_count: number;
  auto_portfolio_count: number;
  recommendation_replay_hit_rate?: number;
  replay_validation_status?: string;
  replay_validation_note?: string | null;
  rule_pass_rate: number;
  run_health: OperationsRunHealthView;
  research_validation: OperationsResearchValidationView;
  launch_readiness: OperationsLaunchReadinessView;
}

export interface OperationsRunHealthView {
  status: string;
  note?: string | null;
  market_data_timeframe: string;
  last_market_data_at?: string | null;
  data_latency_seconds?: number | null;
  refresh_cooldown_minutes: number;
  intraday_source_status: string;
}

export interface OperationsResearchValidationView {
  status: string;
  note?: string | null;
  recommendation_contract_status: string;
  benchmark_status: string;
  benchmark_note?: string | null;
  replay_validation_status: string;
  replay_validation_note?: string | null;
  replay_sample_count: number;
  verified_replay_count: number;
  synthetic_replay_count: number;
  manifest_bound_count: number;
  metrics_artifact_count: number;
  artifact_sample_count: number;
  replay_artifact_bound_count: number;
  replay_artifact_manifest_count: number;
  replay_artifact_nonverified_count: number;
  replay_artifact_backed_projection_count: number;
  replay_migration_placeholder_count: number;
  portfolio_backtest_bound_count: number;
  portfolio_backtest_manifest_count: number;
  portfolio_backtest_verified_count: number;
  portfolio_backtest_pending_rebuild_count: number;
  portfolio_backtest_artifact_backed_projection_count: number;
  portfolio_backtest_migration_placeholder_count: number;
}

export interface OperationsLaunchReadinessView {
  status: string;
  note?: string | null;
  blocking_gate_count: number;
  warning_gate_count: number;
  synthetic_fields_present: boolean;
  recommended_next_gate?: string | null;
  rule_pass_rate: number;
}

export interface AccessControlView {
  beta_phase: string;
  auth_mode: string;
  required_header: string;
  allowlist_slots: number;
  active_users: number;
  roles: string[];
  session_ttl_minutes: number;
  audit_log_retention_days: number;
  export_policy: string;
  alerts: string[];
}

export interface RefreshScheduleView {
  scope: string;
  cadence_minutes: number;
  market_delay_minutes: number;
  stale_after_minutes: number;
  trigger: string;
}

export interface RefreshPolicyView {
  market_timezone: string;
  cache_ttl_seconds: number;
  manual_refresh_cooldown_minutes: number;
  schedules: RefreshScheduleView[];
}

export interface PerformanceThresholdView {
  metric: string;
  unit: string;
  target: number;
  observed: number;
  status: string;
  note: string;
}

export interface LaunchGateView {
  gate: string;
  threshold: string;
  current_value: string;
  status: string;
}

export interface ManualResearchQueueView {
  generated_at: string;
  focus_symbol?: string | null;
  counts: Record<string, number>;
  focus_request?: ManualResearchRequestView | null;
  recent_items: ManualResearchRequestView[];
}

export interface OperationsTodayAtAGlanceView {
  latest_refresh_at?: string | null;
  refresh_status?: string | null;
  data_quality_status?: string | null;
  abnormal_symbol_count?: number;
  event_analysis_count?: number;
  manual_queue_counts?: Record<string, number>;
  top_warning_gate?: string | null;
  top_warning_status?: string | null;
  recommendation_replay_count?: number;
  active_watchlist_count?: number;
  target_watchlist_count?: number;
  missing_target_symbols?: string[];
  replacement_candidates?: string[];
  research_validation_status?: string | null;
  summary_items?: string[];
}

export interface ImprovementSuggestionReview {
  reviewer: string;
  stance: string;
  confidence: number;
  main_reason: string;
  evidence_refs_used: string[];
  missing_evidence: string[];
  implementation_notes: string[];
  red_flags: string[];
  safe_to_plan: boolean;
  safe_to_auto_apply: boolean;
  status: string;
}

export interface ImprovementSuggestionPlan {
  title: string;
  summary: string;
  implementation_steps: string[];
  tests: string[];
  blocked_by: string[];
}

export interface ImprovementSuggestionControlTask {
  id: string;
  title: string;
  status: string;
  model: string;
  project_id: string;
  plan_mode: boolean;
  api_base?: string | null;
  raw_status?: string | null;
  publish_status?: string | null;
  publish_verified?: boolean;
  status_source?: string | null;
  status_stale?: boolean;
  updated_at?: string | null;
  workflow_gates?: Record<string, any> | null;
}

export interface ImprovementSuggestionView {
  suggestion_id: string;
  source_type: string;
  source_ref: string;
  symbol?: string | null;
  category: string;
  claim: string;
  proposed_change: string;
  evidence_refs: string[];
  status: string;
  created_at: string;
  model_consensus?: string;
  evidence_status?: string;
  final_confidence?: string;
  recommended_action?: string;
  decision_reason?: string;
  generated_plan?: ImprovementSuggestionPlan;
  reviews?: Record<string, ImprovementSuggestionReview>;
  control_plane_task?: ImprovementSuggestionControlTask;
}

export interface ImprovementSuggestionsPayload {
  generated_at?: string | null;
  status?: string | null;
  snapshot_file?: string | null;
  window_days?: number;
  model_status: Record<string, string>;
  summary: Record<string, any>;
  top_suggestions?: ImprovementSuggestionView[];
  suggestions?: ImprovementSuggestionView[];
}

export interface OperationsDashboardResponse {
  overview: OperationsOverviewView;
  market_data_timeframe: string;
  last_market_data_at?: string | null;
  data_latency_seconds?: number | null;
  intraday_source_status: IntradaySourceStatusView;
  portfolios: PortfolioSummaryView[];
  recommendation_replay: RecommendationReplayView[];
  access_control: AccessControlView;
  refresh_policy: RefreshPolicyView;
  performance_thresholds: PerformanceThresholdView[];
  launch_gates: LaunchGateView[];
  manual_research_queue: ManualResearchQueueView;
  simulation_workspace?: SimulationWorkspaceResponse | null;
  data_quality_summary?: Record<string, any>;
  factor_observation_summary?: Record<string, any>;
  sector_exposure?: Record<string, any>;
  benchmark_context?: Record<string, any>;
  policy_governance?: Record<string, any>;
  today_at_a_glance?: OperationsTodayAtAGlanceView;
  improvement_suggestions?: ImprovementSuggestionsPayload;
}

File: src/ashare_evidence/schemas/operations.py
"""O p e r a t i o n s domain schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from ashare_evidence.contract_status import STATUS_PENDING_REBUILD

from .portfolio import PortfolioSummaryView
from .research import ManualResearchRequestView
from .simulation import SimulationWorkspaceResponse


class PricePointView(BaseModel):
    observed_at: datetime
    open_price: float
    high_price: float
    low_price: float
    close_price: float
    volume: float


class RecommendationReplayView(BaseModel):
    source: str | None = None
    source_classification: str | None = None
    artifact_type: str | None = None
    artifact_id: str | None = None
    manifest_id: str | None = None
    recommendation_id: int
    recommendation_key: str | None = None
    symbol: str
    stock_name: str
    direction: str
    generated_at: datetime
    label_definition: str
    review_window_definition: str
    entry_time: datetime
    exit_time: datetime
    review_window_days: int | None = None
    stock_return: float
    benchmark_return: float
    excess_return: float
    max_favorable_excursion: float
    max_adverse_excursion: float
    benchmark_definition: str | None = None
    benchmark_source: str | None = None
    validation_mode: str | None = None
    hit_definition: str
    hit_status: str
    validation_status: str = STATUS_PENDING_REBUILD
    validation_note: str | None = None
    summary: str
    followed_by_portfolios: list[str] = Field(default_factory=list)


class OperationsRunHealthView(BaseModel):
    status: str
    note: str | None = None
    market_data_timeframe: str
    last_market_data_at: datetime | None = None
    data_latency_seconds: int | None = None
    refresh_cooldown_minutes: int
    intraday_source_status: str


class ScheduledRefreshComponentView(BaseModel):
    slot: str
    label: str
    status: str
    status_label: str
    message: str
    started_at: str | None = None
    completed_at: str | None = None
    failed_at: str | None = None
    deferred_at: str | None = None
    exit_code: int | None = None
    state_updated_at: str | None = None


class ScheduledRefreshStatusView(BaseModel):
    status: str
    label: str
    message: str
    target_date: str
    slot: str
    scheduled_time: str
    started_at: str | None = None
    completed_at: str | None = None
    failed_at: str | None = None
    deferred_at: str | None = None
    exit_code: int | None = None
    pid: int | None = None
    state_updated_at: str | None = None
    next_action: str | None = None
    components: list[ScheduledRefreshComponentView] = Field(default_factory=list)


class Phase5HorizonSelectionSummaryView(BaseModel):
    approval_state: str
    candidate_frontier: list[int] = Field(default_factory=list)
    lagging_horizons: list[int] = Field(default_factory=list)
    included_record_count: int = 0
    included_as_of_date_count: int = 0
    artifact_id: str | None = None
    artifact_available: bool = False
    note: str | None = None


class Phase5HoldingPolicyStudySummaryView(BaseModel):
    approval_state: str
    included_portfolio_count: int = 0
    mean_turnover: float | None = None
    mean_annualized_excess_return_after_baseline_cost: float | None = None
    gate_status: str | None = None
    governance_status: str | None = None
    governance_action: str | None = None
    redesign_status: str | None = None
    redesign_focus_areas: list[str] = Field(default_factory=list)
    redesign_triggered_signal_ids: list[str] = Field(default_factory=list)
    redesign_primary_experiment_ids: list[str] = Field(default_factory=list)
    failing_gate_ids: list[str] = Field(default_factory=list)
    artifact_id: str | None = None
    artifact_available: bool = False
    note: str | None = None


class OperationsResearchValidationView(BaseModel):
    status: str = STATUS_PENDING_REBUILD
    note: str | None = None
    recommendation_contract_status: str = STATUS_PENDING_REBUILD
    benchmark_status: str = STATUS_PENDING_REBUILD
    benchmark_note: str | None = None
    replay_validation_status: str = STATUS_PENDING_REBUILD
    replay_validation_note: str | None = None
    replay_sample_count: int = 0
    verified_replay_count: int = 0
    synthetic_replay_count: int = 0
    manifest_bound_count: int = 0
    metrics_artifact_count: int = 0
    artifact_sample_count: int = 0
    replay_artifact_bound_count: int = 0
    replay_artifact_manifest_count: int = 0
    replay_artifact_nonverified_count: int = 0
    replay_artifact_backed_projection_count: int = 0
    replay_migration_placeholder_count: int = 0
    portfolio_backtest_bound_count: int = 0
    portfolio_backtest_manifest_count: int = 0
    portfolio_backtest_verified_count: int = 0
    portfolio_backtest_pending_rebuild_count: int = 0
    portfolio_backtest_artifact_backed_projection_count: int = 0
    portfolio_backtest_migration_placeholder_count: int = 0
    phase5_horizon_selection: Phase5HorizonSelectionSummaryView | None = None
    phase5_holding_policy_study: Phase5HoldingPolicyStudySummaryView | None = None


class OperationsLaunchReadinessView(BaseModel):
    status: str
    note: str | None = None
    blocking_gate_count: int = 0
    warning_gate_count: int = 0
    synthetic_fields_present: bool = False
    recommended_next_gate: str | None = None
    rule_pass_rate: float = 0.0


class OperationsOverviewView(BaseModel):
    generated_at: datetime
    beta_readiness: str | None = None
    manual_portfolio_count: int
    auto_portfolio_count: int
    recommendation_replay_hit_rate: float | None = None
    replay_validation_status: str | None = STATUS_PENDING_REBUILD
    replay_validation_note: str | None = None
    rule_pass_rate: float
    run_health: OperationsRunHealthView
    research_validation: OperationsResearchValidationView
    launch_readiness: OperationsLaunchReadinessView


class IntradaySourceStatusView(BaseModel):
    status: str
    provider_name: str | None = None
    provider_label: str | None = None
    source_kind: str
    timeframe: str
    decision_interval_seconds: int
    market_data_interval_seconds: int
    symbol_count: int
    last_success_at: str | None = None
    latest_market_data_at: str | None = None
    data_latency_seconds: int | None = None
    fallback_used: bool = False
    stale: bool = False
    message: str | None = None


class AccessControlView(BaseModel):
    beta_phase: str
    auth_mode: str
    required_header: str
    allowlist_slots: int
    active_users: int
    roles: list[str] = Field(default_factory=list)
    session_ttl_minutes: int
    audit_log_retention_days: int
    export_policy: str
    alerts: list[str] = Field(default_factory=list)


class RefreshScheduleView(BaseModel):
    scope: str
    cadence_minutes: int
    market_delay_minutes: int
    stale_after_minutes: int
    trigger: str


class RefreshPolicyView(BaseModel):
    market_timezone: str
    cache_ttl_seconds: int
    manual_refresh_cooldown_minutes: int
    schedules: list[RefreshScheduleView] = Field(default_factory=list)


class PerformanceThresholdView(BaseModel):
    metric: str
    unit: str
    target: float
    observed: float
    status: str
    note: str


class LaunchGateView(BaseModel):
    gate: str
    threshold: str
    current_value: str
    status: str


class ManualResearchQueueView(BaseModel):
    generated_at: datetime
    focus_symbol: str | None = None
    counts: dict[str, int] = Field(default_factory=dict)
    focus_request: ManualResearchRequestView | None = None
    recent_items: list[ManualResearchRequestView] = Field(default_factory=list)


class OperationsDashboardResponse(BaseModel):
    overview: OperationsOverviewView
    market_data_timeframe: str
    last_market_data_at: datetime | None = None
    data_latency_seconds: int | None = None
    intraday_source_status: IntradaySourceStatusView
    portfolios: list[PortfolioSummaryView] = Field(default_factory=list)
    recommendation_replay: list[RecommendationReplayView] = Field(default_factory=list)
    access_control: AccessControlView
    refresh_policy: RefreshPolicyView
    performance_thresholds: list[PerformanceThresholdView] = Field(default_factory=list)
    launch_gates: list[LaunchGateView] = Field(default_factory=list)
    manual_research_queue: ManualResearchQueueView
    simulation_workspace: SimulationWorkspaceResponse | None = None
    data_quality_summary: dict[str, Any] = Field(default_factory=dict)
    factor_observation_summary: dict[str, Any] = Field(default_factory=dict)
    sector_exposure: dict[str, Any] = Field(default_factory=dict)
    benchmark_context: dict[str, Any] = Field(default_factory=dict)
    today_at_a_glance: dict[str, Any] = Field(default_factory=dict)
